plugins/util/portmap/portmap.go

```diff
@@ -0,0 +1,223 @@
+
+	var ipt *iptables.IPTables
+	var err error
+
+	if isV6 {
+		ipt, err = iptables.NewWithProtocol(iptables.ProtocolIPv6)
+	} else {
+		ipt, err = iptables.NewWithProtocol(iptables.ProtocolIPv4)
+	}
+	if err != nil {
+		return fmt.Errorf("failed to open iptables:", err)
+	}
+
+	dnatChain := genDnatChain(config.Name, config.ContainerID)
+	dnatRules := dnatRules(config.RuntimeConfig.PortMaps, containerIP)
+	if err := dnatChain.setup(ipt, dnatRules); err != nil {
+		return fmt.Errorf("unable to setup DNAT", err)
+	}
+
+	if enableSNAT {
+		snatChain := genSnatChain(config.Name, config.ContainerID, isV6)
+		snatRules := snatRules(config.RuntimeConfig.PortMaps, containerIP)
+		if err := snatChain.setup(ipt, snatRules); err != nil {
+			return fmt.Errorf("unable to setup SNAT", err)
+		}
+		if !isV6 {
+			hostIfName := getRoutableHostIF(config.PrevResult)
+			if hostIfName != "" {
+				if err := enableLocalnetRouting(hostIfName); err != nil {
+					return fmt.Errorf("unable to enable route_localnet: %v", err)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func genDnatChain(netName, containerID string) chain {
+	name := utils.FormatChainName("DN-", netName, containerID)
+	comment := fmt.Sprintf(`dnat name: "%s" id: "%s"`, netName, containerID)
+
+	return chain{
+		name: name,
+		entryRule: []string{
+			"-m", "addrtype",
+			"--dst-type", "LOCAL",
+			"-m", "comment",
+			"--comment", comment,
+			"-j", name},
+		entryChains: []string{"PREROUTING", "OUTPUT"},
+	}
+}
+
+// dnatRules generates the destination NAT rules, one per port, to direct
+// traffic from hostip:hostport to podip:podport
+func dnatRules(entries []PortMapEntry, containerIP net.IP) [][]string {
+	out := make([][]string, 0, len(entries))
+	for _, entry := range entries {
+		out = append(out, []string{
+			"-p", entry.Protocol,
+			"--dport", strconv.Itoa(entry.HostPort),
+			"-j", "DNAT", "--to-destination",
+			fmtIpPort(containerIP, entry.ContainerPort),
+		})
+	}
+	return out
+}
+func genSnatChain(netName, containerID string, isV6 bool) chain {
+	name := utils.FormatChainName("SN-", netName, containerID)
+	comment := fmt.Sprintf(`snat name: "%s" id: "%s"`, netName, containerID)
+
+	if isV6 {
+		return chain{
+			name: name,
+			entryRule: []string{
+				"-s", "::1",
+				"!", "-d", "::1",
+				"-m", "comment",
+				"--comment", comment,
+				"-j", name},
+			entryChains: []string{"POSTROUTING"},
+		}
+	} else {
+		return chain{
+			name: name,
+			entryRule: []string{
+				"-s", "127.0.0.1",
+				"!", "-d", "127.0.0.1",
+				"-m", "comment",
+				"--comment", comment,
+				"-j", name},
+			entryChains: []string{"POSTROUTING"},
+		}
+	}
+}
+
+// snatRules sets up masquerading for connections to localhost:hostport,
+// rewriting the source so that returning packets are correct.
+func snatRules(entries []PortMapEntry, containerIP net.IP) [][]string {
+	isV6 := (containerIP.To4() == nil)
+	var localhostIP string
+	if isV6 {
+		localhostIP = "::1"
+	} else {
+		localhostIP = "127.0.0.1"
```

Comment:
There's already a couple of these, so why not make a small helper function that takes a bool (isV4 or something) returns the right address?  Then you can use it in genSnatChain() too and save some LoC.